### PR TITLE
Homework02

### PR DESCRIPTION
**Pull request by ArzamastsevOleksandr**

### CODE REVIEW

File: homework02/model/User.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+'use strict';
+
+class User {
+    constructor(id, login, password, age, isDeleted) {
```
Comment (DiannaDea) [inline homework02/model/User.js]: TODO: remove `isDeleted` from User constructor, set `this.isDeleted=false` by default.

File: homework02/in-memory-db/db.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
+'use strict';
+
+const User = require('../model/User');
+
+class DB {
+    constructor() {
+        this.users = new Map([
```
Comment (DiannaDea) [inline homework02/in-memory-db/db.js]: 👍🏻

File: homework02/in-memory-db/db.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        ]);
+    }
+
+    fetchAllUsers() {
+        return [...this.users.values()]
+            .filter(user => !user.isDeleted);
+    }
+
+    getById(id) {
+        const userToDelete = this.users.get(id);
```
Comment (DiannaDea) [inline homework02/in-memory-db/db.js]: TODO: change variable name `userToDelete`. It's just a user you are trying to get.

File: homework02/in-memory-db/db.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+        return userToDelete.isDeleted ? undefined : userToDelete;
```
Comment (DiannaDea) [inline homework02/in-memory-db/db.js]: TODO: better to use `null` instead of `undefined`.

File: homework02/in-memory-db/db.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+    save(user) {
+        this.users.set(user.id, user);
+    }
+
+    getAutoSuggestedUsers(loginSubstring, limit) {
+        const comparator = (u1, u2) => u1.login.localeCompare(u2.login);
+
+        return [...this.users.values()]
+            .slice()
```
Comment (DiannaDea) [inline homework02/in-memory-db/db.js]: Why do we need `slice()`?

File: homework02/in-memory-db/db.js [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+            .filter(u => u.login.includes(loginSubstring))
+            .sort(comparator)
+            .splice(0, limit);
```
Comment (DiannaDea) [inline homework02/in-memory-db/db.js]: TODO: use `.slice(0, limit);`

`splice()` is used mostly to insert or replace elements in array. We can use this method for `slice` functionality, but there is no need doing it.
`slice()` is used to get a part of array starting from position A to position B.

File: homework02/controllers/user-controller.js [hunk 7]
```diff
@@ -0,0 +1,114 @@
+'use strict';
+
+const DB = require('../in-memory-db/db');
+const User = require('../model/User');
+
+const db = new DB();
+
+exports.greeting = (req, res) => {
+    res.render('index', {
```
Comment (DiannaDea) [inline homework02/controllers/user-controller.js]: TODO: remove rendering page. You can create route e.g `/health-check` to check whether your API is working or not.

File: homework02/controllers/user-controller.js [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+
+exports.findAll = (req, res) => {
+    res.setHeader('Content-Type', 'application/json');
+    res.status(200).end(JSON.stringify(db.fetchAllUsers(), null, 3));
+};
+
+exports.findById = (req, res) => {
+    const result = db.getById(req.params.id);
+    if (!result) {
+        res.status(400).send({
```
Comment (DiannaDea) [inline homework02/controllers/user-controller.js]: TODO:  use status code 404. 400 is used for bad client requests, e.g validation errors

File: homework02/controllers/user-controller.js [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+            message: `User not found with id [${req.params.id}]`
+        });
+    }
+    res.setHeader('Content-Type', 'application/json');
+    res.end(JSON.stringify(result, null, 3));
```
Comment (DiannaDea) [inline homework02/controllers/user-controller.js]: JSON.stringify(result) is enough. I think there is no need to pass extra params here

File: homework02/controllers/user-controller.js [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+    const params = req.params;
+
+    res.setHeader('Content-Type', 'application/json');
+    res.end(JSON.stringify(db.getAutoSuggestedUsers(params.loginSubstring, params.limit), null, 3));
+};
+
+exports.create = (req, res) => {
+    const body = req.body;
+
+    if (!body.id) {
```
Comment (DiannaDea) [inline homework02/controllers/user-controller.js]: TODO: we shouldn't pass `userId` from client side. It should be generated on backend side

File: homework02/controllers/user-controller.js [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+    userToUpdate.isDeleted = body.isDeleted;
+}
+
+exports.updateById = (req, res) => {
+    const params = req.params;
+
+    const userToUpdate = db.getById(params.id);
+    if (!userToUpdate) {
+        return res.status(400).send({
+            errorMessage: `No user found with id [${params.id}]`
```
Comment (DiannaDea) [inline homework02/controllers/user-controller.js]: Status code 404

File: homework02/controllers/user-controller.js [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+
+    return res.status(200).send({
+        'result': userToUpdate
+    });
+};
+
+exports.deleteById = (req, res) => {
+    const body = req.body;
+
+    if (!body.id) {
```
Comment (DiannaDea) [inline homework02/controllers/user-controller.js]: TODO: user id should be in `request.params`, but not `request.body`
Route for deleting user: DELETE /users/:id

File: homework02/controllers/user-controller.js [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+        return res.status(400).send({
+            message: 'Can not delete a user without id!'
+        });
+    }
+
+    const userToDelete = db.getById(body.id);
+
+    if (!userToDelete) {
+        return res.status(400).send({
```
Comment (DiannaDea) [inline homework02/controllers/user-controller.js]: 404

File: homework02/controllers/user-controller.js [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+            message: `User with id [${body.id}] does not exist!`
+        });
+    }
+
+    userToDelete.setIsDeleted('T');
```
Comment (DiannaDea) [inline homework02/controllers/user-controller.js]: TODO: let's use just boolean values

File: homework02/controllers/user-controller.js [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+
+    return res.status(200).send({
+        message: `User with id [${body.id}] is successfully deleted.`,
+        user: userToDelete
```
Comment (DiannaDea) [inline homework02/controllers/user-controller.js]: TODO: no need to send deleted user

File: homework02/package.json [hunk 16]
```diff
@@ -0,0 +1,24 @@
+{
+  "name": "homework02",
+  "version": "0.0.0",
+  "private": true,
+  "scripts": {
+    "babel": "babel --presets es2015 bin/www.js -o transpiled.js",
+    "start": "node bin/www.js"
+  },
+  "dependencies": {
+    "cookie-parser": "1.4.4",
```
Comment (DiannaDea) [inline homework02/package.json]: Remove

File: homework02/package.json [hunk 17]
```diff
@@ -0,0 +1,24 @@
...
+    "debug": "2.6.9",
+    "express": "4.16.1",
+    "http-errors": "1.6.3",
+    "joi": "14.3.1",
+    "morgan": "1.9.1",
+    "pug": "2.0.4"
```
Comment (DiannaDea) [inline homework02/package.json]: Remove

File: homework02/public/stylesheets/style.css [hunk 18]
```diff
@@ -0,0 +1,8 @@
+body {
```
Comment (DiannaDea) [inline homework02/public/stylesheets/style.css]: Remove

File: homework02/routes/user-routes.js [hunk 19]
```diff
@@ -0,0 +1,23 @@
...
+
+    app.get('/users/:id', users.findById);
+
+    app.get('/users/:loginSubstring/:limit', users.autoSuggestionByLogin);
+
+    app.post('/users', validateRequest, users.create);
+
+    app.put('/users/:id', validateRequest, users.updateById);
+
+    app.delete('/users', users.deleteById);
```
Comment (DiannaDea) [inline homework02/routes/user-routes.js]: TODO: correct endpoint /users/:id

File: homework02/routes/user-routes.js [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+module.exports = (app) => {
+    const users = require('../controllers/user-controller');
+
+    app.get('/', users.greeting);
+
+    app.get('/users', users.findAll);
+
+    app.get('/users/:id', users.findById);
+
+    app.get('/users/:loginSubstring/:limit', users.autoSuggestionByLogin);
```
Comment (DiannaDea) [inline homework02/routes/user-routes.js]: TODO: this endpoint is not by REST api rules. The correct endpoint should be: /users?loginSubstring='ldldl'&limit=5. Moreover /users and /users?loginSubstring='ldldl'&limit=5 should be handled as one endpoint. 

So: 

```
users.findAll(req, res) {
  if (req.query.loginSubstring && req.query.limit) { 
         //  call users.autoSuggestionByLogin(loginSubstring, limit)
        // pay attention that we are not calling users.autoSuggestionByLogin() with req, res
   }
  else // call users.findAll()
}
```

File: homework02/schemas/schemas.js [hunk 21]
```diff
@@ -0,0 +1,21 @@
+'use strict';
+
+const Joi = require('joi');
+
+const userId = Joi.string().min(4).max(10);
+const userLogin = Joi.string().alphanum().min(3);
+const userPassword = Joi.string().min(7).alphanum().strict();
+const userAge = Joi.number().integer().min(4).max(130);
+const userIsDeleted = Joi.string().valid('T', 'TRUE').uppercase();
```
Comment (DiannaDea) [inline homework02/schemas/schemas.js]: TODO: boolean value

File: homework02/views/layout.pug [hunk 22]
```diff
@@ -0,0 +1,7 @@
+doctype html
```
Comment (DiannaDea) [inline homework02/views/layout.pug]: Remove

File: homework02/views/index.pug [hunk 23]
```diff
@@ -0,0 +1,5 @@
+extends layout
```
Comment (DiannaDea) [inline homework02/views/index.pug]: Remove

File: homework02/views/error.pug [hunk 24]
```diff
@@ -0,0 +1,6 @@
+extends layout
```
Comment (DiannaDea) [inline homework02/views/error.pug]: Remove

File: homework02/validator/validator.js [hunk 25]
```diff
@@ -0,0 +1,52 @@
+const lodash = require('lodash');
+const Joi = require('joi');
+const Schemas = require('../schemas/schemas');
+
+module.exports = (useJoiError = false) => {
+    const _useJoiError = lodash.isBoolean(useJoiError) && useJoiError;
+
+    const supportedMethods = ['post', 'put'];
```
Comment (DiannaDea) [inline homework02/validator/validator.js]: TODO: move out of function scope. It's static variable

File: homework02/validator/validator.js [hunk 26]
```diff
@@ -0,0 +1,52 @@
...
+
+    const validationOptions = {
```
Comment (DiannaDea) [inline homework02/validator/validator.js]: TODO: move out of function scope. It's static variable

File: homework02/validator/validator.js [hunk 27]
```diff
@@ -0,0 +1,52 @@
...
+                            error: {
+                                original: err._object,
+                                details: lodash.map(err.details, ({ message, type }) => ({
+                                    message: message.replace(/['"]/g, ''),
+                                    type
+                                }))
+                            }
+                        };
+
+                        const CustomError = {
```
Comment (DiannaDea) [inline homework02/validator/validator.js]: TODO: move out of function scope. It's static variable

File: homework02/validator/validator.js [hunk 28]
```diff
@@ -0,0 +1,52 @@
...
+                            status: 'failed',
+                            error: 'Invalid request data. Please review request and try again.'
+                        };
+
+                        res.status(422).json(_useJoiError ? JoiError : CustomError);
+                    } else {
+                        req.body = data;
```
Comment (DiannaDea) [inline homework02/validator/validator.js]: TODO: no need to rewrite `req.body`

**Comment by DiannaDea** [comment]
General feedback:

**1. Remove views.** 

Since the goal of this homework is to create REST api, we don't need views. Moreover, I've never seen node.js applications with client as views on backend side. 

**2. CRUD.** 

Simple CRUD should have the following routes:

- POST /users - create user
- PUT /users/:id - update all user fields
- PATCH /users/:id - partially update user fields
- DELETE /users/:id - delete user
- GET /users/:id - get singe user
- GET /users - get all users. Please pay attention, that if we want to sort, filter, get only particular number or all users, we still use this endpoint. 

**The logic:**
Params after `?` are query params. If we don't have query params, we return all users. If we have query params, we return filtered, sorted, etc users

**Examples:**
GET /users?limit=100&offset=20 - example of pagination
GET /users?login='lalla'&limit=20 - as in your homework filter users by login

Status codes:
**200** - success
**201** - successfully created
**400** - bad client request
**404**- endpoint not found or e.g user not found
**422** - validation error
**500** - internal server error

3. Shared response template.

The JSON structure of every response you send from server should be the same.
Example:

Success request:
```
{
  error: null,
  data: { // any information you want to send }
}
```
Failed request:

```
{
  error: "Error message",
  data: null
}
```

4. Use middlewares.

Example: almost in every request you need to check if user with such id exists or not. It is a good practice to create middleware function that checks it to avoid code duplication.

5. Generating ids

Ids are not passed in request body, only in request params. If we create user, we generate id on server side.

File: homework02/controllers/user-controller.js (hunk 12 above)
Comment (ArzamastsevOleksandr) [reply homework02/controllers/user-controller.js]: Done.

File: homework02/controllers/user-controller.js (hunk 13 above)
Comment (ArzamastsevOleksandr) [reply homework02/controllers/user-controller.js]: Done.

File: homework02/controllers/user-controller.js (hunk 15 above)
Comment (ArzamastsevOleksandr) [reply homework02/controllers/user-controller.js]: Understood, removed this logic.

File: homework02/routes/user-routes.js (hunk 19 above)
Comment (ArzamastsevOleksandr) [reply homework02/routes/user-routes.js]: Done.

**Pull request merged**